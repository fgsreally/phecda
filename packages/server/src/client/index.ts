import type { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios'
import type { MergeType, RequestType, ResOrErr, Transform } from '../types'
interface RequestArgs {
  body: Record<string, any>
  query: Record<string, string>
  params: Record<string, string>
  realParam: string
  method: RequestType
  url: string
  name: string
}
type MergedReqArg = Pick<RequestArgs, 'body' | 'query' | 'params'>
export function toReq(arg: RequestArgs) {
  const { body, query, realParam, method, url } = arg
  return { method, url, body, query: Object.keys(query).length > 0 ? `?${Object.entries(query).map(([k, v]) => `${k}=${v}`).join('&')}` : '', params: realParam }
}
// @ts-expect-error misdirction
export const merge: MergeType = (...args: RequestArgs[]) => {
  const ret = {} as Record<string, MergedReqArg>
  for (const i of args) {
    const { name, body, query, params } = i
    ret[name] = { body, query, params }
  }

  return ret
}
export type RequestMethod = <F extends (...args: any[]) => any >(fn: F, args: Parameters<F>) => Promise<ReturnType<F>>

export function createReq(instance: AxiosInstance): <R>(arg: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<Awaited<R>> > {
  // @ts-expect-error methods without route decorator won't send request
  return (arg: any, config?: AxiosRequestConfig) => {
    const { url, params, query, body, method } = toReq(arg as RequestArgs)
    if (!method) {
      console.warn('methods without route decorator won\'t send request')
      return
    }

    const ret = [`${url}${params}${query}`] as any[]
    body && ret.push(body)
    config && ret.push(config)
    // @ts-expect-error misdirction

    return instance[method](...ret)
  }
}

export function createMergeReq(instance: AxiosInstance, key = '/__PHECDA_SERVER__'): < R extends any[]>(args: R, config?: AxiosRequestConfig) => Promise<AxiosResponse<ResOrErr<R>>> {
  // @ts-expect-error misdirction

  return (args: Record<string, MergedReqArg>, config?: AxiosRequestConfig) => {
    return instance.post(key, args, config)
  }
}

export function P<C extends new (...args: any) => any>(Model: C): Transform<C> {
  return new Model()
}
