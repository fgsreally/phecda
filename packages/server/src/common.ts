export const MERGE_SYMBOL = '__PS_MERGE__'
export const UNMOUNT_SYMBOL = '__PS_UNMOUNT__'
export const MODULE_SYMBOL = '__PS_MODULE__'
export const META_SYMBOL = '__PS_META__'
export const APP_SYMBOL = '__PS__'
export const IS_DEV=process.env.NODE_ENV === 'development'