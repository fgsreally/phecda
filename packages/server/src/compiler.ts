import type { RequestType } from './types'

export class Pcompiler {
  content = ''
  classMap: Record<string, { [key: string]: string }> = {}
  constructor() { }

  getContent() {
    let content = ''

    for (const name in this.classMap) {
      content += `
        export class ${name}{
            ${Object.values(this.classMap[name]).reduce((p, c) => p + c)}
            }`
    }
    return content
  }

  addMethod(className: string, methodName: string, route = '', requestType: RequestType | '' = '', params: { type: string; key: string; index: number }[] = []) {
    const url = route.replace(/\/\:([^\/]*)/g, '')
    if (!this.classMap[className])
      this.classMap[className] = {}
    this.classMap[className][methodName] = `
    ${methodName}(${genParams(params)}){
const ret={name:"${className}-${methodName}",body:{},query:{},params:{},realParam:'',method:"${requestType}",url:"${url}"}
${params.reduce((p, c) => `${p}ret.${c.type}.${c.key}=arg${c.index}\n${c.type === 'params' ? `ret.realParam+='/'+arg${c.index}\n` : ''}`, '')}
return ret
    }
    `
  }
}

// function genExpression(type: string, key: string, index: number) {
//   switch (type) {
//     case 'params':
//       return `ret.params+='/'+arg${index}`
//     case 'query':
//       return `ret.query+=ret.query?"&${key}="+arg${index}:"?${key}="+arg${index}`
//     case 'body':
//       return `ret.body[${key}]=arg${index}`
//   }
// }

function genParams(decorators: any[]) {
  let index = 0
  return decorators.reduce((p) => {
    return `${`${p}arg${index++}`},`
  }, '')
}
