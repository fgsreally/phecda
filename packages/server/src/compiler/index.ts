export * from './rpc'
export * from './http'
