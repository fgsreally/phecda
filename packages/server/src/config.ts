export const Pconfig = {
  rabbitmq: {
    guard: false,
    interceptor: false,
  },
}
