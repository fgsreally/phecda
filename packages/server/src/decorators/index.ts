import { SHARE_KEY, setState, setVar } from 'phecda-core'

export function Header(name: string, value: string) {
  return (target: any, k: PropertyKey) => {
    setVar(target, k)
    const state = target._namespace.__STATE_NAMESPACE__.get(k) || {}
    if (!state.header)
      state.header = {}

    state.header[name] = value
    setState(target, k, state)
  }
}

export function Define(key: string, value: any) {
  return (target: any, k?: PropertyKey) => {
    if (!k) {
      k = SHARE_KEY
      target = target.prototype
    }

    setVar(target, k)
    const state = target._namespace.__STATE_NAMESPACE__.get(k) || {}
    if (!state.define)
      state.define = {}

    state.define[key] = value
    setState(target, k, state)
  }
}

export * from './param'
export * from './route'
export * from './aop'
