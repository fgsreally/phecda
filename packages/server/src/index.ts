import 'reflect-metadata'

type Construct<T = any> = new (...args: Array<any>) => T

export function Factory<T>(Module: Construct<T>) {
  const moduleMap = new Map<Construct, InstanceType<Construct>>()
  const app = buildNestModule(Module, moduleMap) as InstanceType<Construct<T>>
  return { app, moduleMap }
}

function buildNestModule(Module: Construct, map: Map<Construct, InstanceType<Construct>>) {
  const paramtypes = getParamtypes(Module) as Construct[]
  let instance: InstanceType<Construct>
  if (map.has(Module)) {
    instance = map.get(Module)
    if (!instance)
      throw new Error(`exist Circular Module dep--${Module}`)
    return instance
  }
  map.set(Module, undefined)
  if (paramtypes) {
    instance = new Module(...paramtypes.map(item =>
      buildNestModule(item, map),
    ))
  }
  else {
    instance = new Module()
  }
  map.set(Module, instance)

  return instance
}

function getParamtypes(Module: Construct) {
  return Reflect.getMetadata('design:paramtypes', Module)
}
