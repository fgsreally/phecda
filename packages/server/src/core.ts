import 'reflect-metadata'
import fs from 'fs'
import EventEmitter from 'node:events'
import type { Phecda } from 'phecda-core'
import { Empty, getExposeKey, getHandler, getState, injectProperty, isPhecda, registerAsync } from 'phecda-core'
import Debug from 'debug'
import pc from 'picocolors'
import type { Construct, Emitter, P } from './types'
import { Meta } from './meta'
import { log } from './utils'
import { IS_DEV, UNMOUNT_SYMBOL } from './common'
import { generateHTTPCode, generateRPCCode } from './compiler'
export function Injectable() {
  return (target: any) => Empty(target)
}
const debug = Debug('phecda-server')
// TODO: support both emitter types and origin emitter type in future
export const emitter: Emitter = new EventEmitter() as any

export async function Factory(Modules: (new (...args: any) => any)[], opts: {

  // HTTP generate code path
  http?: string
  // rpc generate code path
  rpc?: string
} = {}) {
  const moduleMap = new Map<string, InstanceType<Construct>>()
  const meta: Meta[] = []
  const constructorMap = new Map()

  // only work for warn
  const constructorSet = new WeakSet()
  const moduleGraph = new Map<string, Set<string>>()
  const { http, rpc } = opts
  injectProperty('watcher', ({ eventName, instance, key, options }: { eventName: string; instance: any; key: string; options?: { once: boolean } }) => {
    const fn = typeof instance[key] === 'function' ? instance[key].bind(instance) : (v: any) => instance[key] = v


    // work for hmr
    instance[UNMOUNT_SYMBOL]?.push(()=>{
      emitter.off(eventName as any,fn)
    })

    if (options?.once)
      (emitter as any).once(eventName, fn)

    else
      (emitter as any).on(eventName, fn)
  })

  async function update(Module: Construct) {
    const tag = Module.prototype?.__TAG__ || Module.name
    if (!moduleMap.has(tag))
      return
    debug(`update module "${tag}"`)

    const instance = moduleMap.get(tag)

    if (instance?.[UNMOUNT_SYMBOL]) {
      for (const cb of instance[UNMOUNT_SYMBOL])
        await cb()
    }
    log(`Module ${pc.yellow(`[${tag}]`)} unmount`)
    moduleMap.delete(tag)
    constructorMap.delete(tag)
    for (let i = meta.length - 1; i >= 0; i--) {
      if (meta[i].data.tag === tag)
        meta.splice(i, 1)
    }

    const { instance: newModule } = await buildNestModule(Module)
    if (moduleGraph.has(tag)) {
      [...moduleGraph.get(tag)!].forEach((tag) => {
        const module = moduleMap.get(tag)
        for (const key in module) {
          if (module[key] === instance)
            module[key] = newModule
        }
      })
    }

    moduleMap.set(tag, newModule)
  }
  async function buildNestModule(Module: Construct) {
    const paramtypes = getParamTypes(Module) as Construct[]
    let instance: InstanceType<Construct>
    const tag = Module.prototype?.__TAG__ || Module.name
    if (moduleMap.has(tag)) {
      instance = moduleMap.get(tag)
      if (!instance)
        throw new Error(`exist Circular-Dependency or Multiple modules with the same name/tag [tag] ${tag}--[module] ${Module}`)

      if (constructorMap.get(tag) !== Module && !constructorSet.has(Module)) {
        constructorSet.add(Module)// a module will only warn once
        log(`Synonym module: Module taged "${tag}" has been loaded before, so phecda-server won't load Module "${Module.name}"`, 'warn')
      }
      return { instance, tag }
    }
    moduleMap.set(tag, undefined)
    if (paramtypes) {
      const paramtypesInstances = [] as any[]
      for (const i in paramtypes) {
        const { instance: sub, tag: subTag } = await buildNestModule(paramtypes[i])
        paramtypesInstances[i] = sub
        if (!moduleGraph.has(subTag))
          moduleGraph.set(subTag, new Set())
        moduleGraph.get(subTag)!.add(tag)
      }
      instance = new Module(...paramtypesInstances)
    }
    else {
      instance = new Module()
    }
    meta.push(...getMetaFromInstance(instance, tag, Module.name))
    await registerAsync(instance)
    moduleMap.set(tag, instance)
    log(`Module ${pc.yellow(`[${tag}]`)} mount"`)
    constructorMap.set(tag, Module)
    return { instance, tag }
  }

  for (const Module of Modules)
    await buildNestModule(Module)

  function writeCode() {
    debug('write code')

    http && fs.promises.writeFile(http, generateHTTPCode(meta.map(item => item.data)))
    rpc && fs.promises.writeFile(rpc, generateRPCCode(meta.map(item => item.data)))
  }

  writeCode()
  if (IS_DEV) {
    // @ts-expect-error globalThis
    globalThis.__PS_HMR__ = async (file: string) => {
      debug(`reload file ${file}`)
      const module = await import(file)
      for (const i in module) {
        if (isPhecda(module[i]))
          await update(module[i])
      }
    }
    // @ts-expect-error globalThis
    globalThis.__PS_WRITEMETA__ = writeCode
  }

  return {
    moduleMap,
    meta,
    constructorMap,
    update,
  }
}

function getMetaFromInstance(instance: Phecda, tag: string, name: string) {
  const vars = getExposeKey(instance).filter(item => item !== '__CLASS')
  const baseState = (getState(instance, '__CLASS') || {}) as P.Meta
  initState(baseState)

  return vars.map((i) => {
    const meta = {} as P.Meta
    const state = (getState(instance, i) || {}) as P.Meta
    initState(state)
    if (state.http) {
      meta.http = {
        route: (baseState.http?.route || '') + (state.http.route),
        type: state.http.type,
      }
    }
    if (baseState.rpc)
      meta.rpc = baseState.rpc
    if (state.rpc) {
      meta.rpc = {
        ...meta.rpc,
        ...state.rpc,
      }
    }

    meta.name = name
    meta.tag = tag
    meta.method = i as string
    const params = [] as any[]
    for (const i of state.params || []) {
      params.unshift(i)
      if (i.index === 0)
        break
    }

    meta.params = params
    meta.define = { ...baseState.define, ...state.define }
    meta.header = { ...baseState.header, ...state.header }
    meta.plugins = [...new Set([...baseState.plugins, ...state.plugins])]
    meta.guards = [...new Set([...baseState.guards, ...state.guards])]
    meta.interceptors = [...new Set([...baseState.interceptors, ...state.interceptors])]

    return new Meta(meta as unknown as P.Meta, getHandler(instance, i), getParamTypes(instance, i as string) || [])
  })
}

function getParamTypes(Module: any, key?: string | symbol) {
  return Reflect.getMetadata('design:paramtypes', Module, key!)
}

function initState(state: any) {
  if (!state.define)
    state.define = {}
  if (!state.header)
    state.header = {}
  if (!state.plugins)
    state.plugins = []
  if (!state.guards)
    state.guards = []
  if (!state.interceptors)
    state.interceptors = []
}
