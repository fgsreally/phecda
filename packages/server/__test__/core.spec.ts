import { describe, expect, expectTypeOf, it } from 'vitest'
import { Factory } from '../src/core'
import { Body, Controller, Get, Query } from '../src/decorators'
import type { Pmeta } from '../src/meta'
describe('Factory ', () => {
  it('Factory will create instance and collect metadata', async () => {
    @Controller('/base')
    class A {
      @Get('/test')
      test(@Query('id') id: string, @Body('name') name: string) {
        return id + name
      }
    }
    const { meta } = await Factory([A])
    const data = meta.map(item => item.data)
    expectTypeOf(meta).items.toEqualTypeOf<Pmeta>()
    expect(data).toMatchSnapshot()
  })

  it('Factory will work using nest class', async () => {
    class Service {
      test() {
        return 'test'
      }
    }

    @Controller('/base')
    class A {
      constructor(public service: Service) {
      }

      @Get('/test')
      test() {
        return this.service.test()
      }
    }
    const { moduleMap } = await Factory([A])
    expect(moduleMap.size).toBe(2)
    expect(moduleMap.get('A').test()).toBe('test')
  })
})
