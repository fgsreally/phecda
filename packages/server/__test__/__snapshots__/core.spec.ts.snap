// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Factory  > Factory will create instance and collect metadata 1`] = `
[
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test",
    "middlewares": [],
    "name": "A",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/base/test",
      "type": "get",
    },
    "tag": "A",
  },
]
`;

exports[`Factory  > In extended case 1`] = `
[
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test2",
    "middlewares": [],
    "name": "B",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/B/testB",
      "type": "get",
    },
    "tag": "B",
  },
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test",
    "middlewares": [],
    "name": "B",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/B/testA",
      "type": "get",
    },
    "tag": "B",
  },
]
`;

exports[`Factory  > multple class extends the same class 1`] = `
[
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test2",
    "middlewares": [],
    "name": "B",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/B/testB",
      "type": "get",
    },
    "tag": "B",
  },
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test",
    "middlewares": [],
    "name": "B",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/B/testA",
      "type": "get",
    },
    "tag": "B",
  },
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test3",
    "middlewares": [],
    "name": "C",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/C/testC",
      "type": "get",
    },
    "tag": "C",
  },
  {
    "define": {},
    "guards": [],
    "header": {},
    "interceptors": [],
    "method": "test",
    "middlewares": [],
    "name": "C",
    "params": [
      {
        "index": 0,
        "key": "id",
        "type": "query",
        "validate": undefined,
      },
      {
        "index": 1,
        "key": "name",
        "type": "body",
        "validate": undefined,
      },
    ],
    "route": {
      "route": "/C/testA",
      "type": "get",
    },
    "tag": "C",
  },
]
`;
